statsmodels/regression/heckman.py

```diff
@@ -0,0 +1,654 @@
+        delta_1d = delta.T[0]
+        
+        Q = rhoHat**2 * ((W.T*delta_1d).dot(Z[self.treated])).dot(step1_varcov).dot((Z[self.treated].T*delta_1d).dot(W))
+        
+        WT_W_inv = np.linalg.inv(W.T.dot(W))
+        WT_R = W.T*(1 - rhoHat**2 * delta_1d)
+        normalized_varcov_all = WT_W_inv.dot(WT_R.dot(W)+Q).dot(WT_W_inv)
+        del WT_W_inv
+        del WT_R
+        
+        del delta_1d
+
+        normalized_varcov = normalized_varcov_all[:-1,:-1]
+
+        varcov_all = sigma2Hat * normalized_varcov_all
+        varcov = varcov_all[:-1,:-1]
+
+        stderr_all = np.sqrt(np.diag(varcov_all))
+        stderr = stderr_all[:-1]
+        stderr_betaHat_inverse_mills = stderr_all[-1]
+
+
+        ## store results
+        results = HeckmanResults(self, params, normalized_varcov, sigma2Hat,
+            select_res=step1res,
+            param_inverse_mills=betaHat_inverse_mills, stderr_inverse_mills=stderr_betaHat_inverse_mills,
+            var_reg_error=sigma2Hat, corr_eqnerrors=rhoHat,
+            method='twostep')
+
+        return results
+
+
+    def _fit_mle(self, start_params_mle=None, method_mle=None, maxiter_mle=None, **kwargs_mle):
+        # get number of X parameters and number of Z parameters
+        Y, X, Z = self.get_datamats()
+        num_xvars = X.shape[1]
+        num_zvars = Z.shape[1]
+        del Y, X, Z
+
+        # let the Heckman two-step parameter estimates be the starting values
+        # of the the optimizer of the Heckman MLE estimate if not specified by user
+        if start_params_mle is None:
+            twostep_res = self._fit_twostep()
+
+            xparams = np.asarray(twostep_res.params)
+            zparams = np.asarray(twostep_res.select_res.params)
+            params_all = np.append(xparams, zparams)
+            params_all = np.append(params_all, twostep_res.corr_eqnerrors)
+            params_all = np.append(params_all, twostep_res.var_reg_error)
+
+            start_params_mle = params_all
+
+        # fit Heckman parameters by MLE
+        results_mle = super(Heckman, self).fit(
+            start_params=start_params_mle, method=method_mle, maxiter=maxiter_mle,
+            **kwargs_mle
+            )
+
+        xbeta_hat = np.asarray(results_mle.params[:num_xvars])  # reg eqn coefs
+        zbeta_hat = np.asarray(results_mle.params[num_xvars:num_xvars+num_zvars])  # selection eqn coefs
+        rho_hat = results_mle.params[-2]
+        sigma2_hat = results_mle.params[-1]
+
+        scale = results_mle.scale
+        #TODO: I think these should be the sandwich estimates, but should confirm
+        xbeta_ncov_hat = results_mle.normalized_cov_params[:num_xvars,:num_xvars]
+        zbeta_ncov_hat = results_mle.normalized_cov_params[
+            num_xvars:(num_xvars+num_zvars),num_xvars:(num_xvars+num_zvars)
+            ]
+
+        rho_var_hat = results_mle.normalized_cov_params[-2,-2] * scale
+        sigma2_var_hat = results_mle.normalized_cov_params[-1,-1] * scale
```

Comment:
When fitting by MLE with the `fit` function of the `LikelihoodModel` class (line 305), the variance of rho and sigma2 are reported as negative (i.e. `rho_var_hat` on line 322 and `sigma2_var_hat` on line 323, but this only happens in the Python 2 builds (they are positive as they should be in the Python 3 builds). Any idea why this might be? This causes a test to fail.